File: pyg4ometry/pygeometry/cpgeant4/solid/TwistedBox.cpp
#include "TwistedBox.h"

CSG* CSGMesh::ConstructTwistedBox(double twistedangle, double pDx, double pDy, double pDz, int refine){
  std::vector<std::vector<double> > vert_crd_up(4,std::vector<double>(3));
  std::vector<std::vector<double> > vert_crd_dn(4,std::vector<double>(3));

  vert_crd_up[0][0] = -pDx; vert_crd_up[0][1] = -pDy; vert_crd_up[0][2] = pDz;
  vert_crd_up[1][0] = pDx; vert_crd_up[1][1] = -pDy; vert_crd_up[1][2] = pDz;
  vert_crd_up[2][0] = pDx; vert_crd_up[2][1] = pDy; vert_crd_up[2][2] = pDz;
  vert_crd_up[3][0] = -pDx; vert_crd_up[3][1] = pDy; vert_crd_up[3][2] = pDz;

  vert_crd_dn[0][0] = -pDx; vert_crd_dn[0][1] = -pDy; vert_crd_dn[0][2] = -pDz;
  vert_crd_dn[1][0] = -pDx; vert_crd_dn[1][1] = pDy; vert_crd_dn[1][2] = -pDz;
  vert_crd_dn[2][0] = pDx; vert_crd_dn[2][1] = pDy; vert_crd_dn[2][2] = -pDz;
  vert_crd_dn[3][0] = pDx; vert_crd_dn[3][1] = -pDy; vert_crd_dn[3][2] = -pDz;

  std::vector<std::vector<double> > vert_crd_rot(4);

  for(int i=0;i<4;i++){
    std::vector<double> vrot(3);
    vrot[0] = vert_crd_up[i][0]*cos(twistedangle) - vert_crd_up[i][1]*sin(twistedangle);
    vrot[1] = vert_crd_up[i][0]*sin(twistedangle) + vert_crd_up[i][1]*cos(twistedangle);
    vrot[2] = vert_crd_up[i][2];
    vert_crd_rot[i] = vrot;
  }

  std::vector<Vertex*> vert_up(4);
  std::vector<Vertex*> vert_dn(4);

  for(int i=0;i<4;i++){
    vert_up[i] = new Vertex(new Vector(vert_crd_rot[i][0],vert_crd_rot[i][1],vert_crd_rot[i][2]),NULL);
    vert_dn[i] = new Vertex(new Vector(vert_crd_dn[i][0],vert_crd_dn[i][1],vert_crd_dn[i][2]),NULL);
  } 

  std::vector<Polygon*> polygons(12);
  //There's almost definitely a less stupid way to do this
  std::vector<Vertex*> v0(3);
  v0[0] = vert_dn[0]; v0[1] = vert_dn[1]; v0[2] = vert_dn[2];
  polygons[0] = new Polygon(v0,NULL);

  std::vector<Vertex*> v1(3);
  v1[0] = vert_dn[2]; v1[1] = vert_dn[3]; v1[2] = vert_dn[0];
  polygons[1] = new Polygon(v1,NULL);
  
  std::vector<Vertex*> v2(3);
  v2[0] = vert_up[0]; v2[1] = vert_up[1]; v2[2] = vert_up[2];
  polygons[2] = new Polygon(v2,NULL);
  
  std::vector<Vertex*> v3(3);
  v3[0] = vert_up[2]; v3[1] = vert_up[3]; v3[2] = vert_up[0];
  polygons[3] = new Polygon(v3,NULL);
  
  std::vector<Vertex*> v4(3);
  v4[0] = vert_dn[0]; v4[1] = vert_dn[3]; v4[2] = vert_up[1];
  polygons[4] = new Polygon(v4,NULL);

  std::vector<Vertex*> v5(3);
  v5[0] = vert_dn[0]; v5[1] = vert_up[1]; v5[2] = vert_up[0];
  polygons[5] = new Polygon(v5,NULL);

  std::vector<Vertex*> v6(3);
  v6[0] = vert_dn[1]; v6[1] = vert_up[2]; v6[2] = vert_dn[2]; 
  polygons[6] = new Polygon(v6,NULL);

  std::vector<Vertex*> v7(3);
  v7[0] = vert_dn[1]; v7[1] = vert_up[3]; v7[2] = vert_up[2];
  polygons[7] = new Polygon(v7,NULL);

  std::vector<Vertex*> v8(3);
  v8[0] = vert_dn[0]; v8[1] = vert_up[0]; v8[2] = vert_up[3];
  polygons[8] = new Polygon(v8,NULL);

  std::vector<Vertex*> v9(3);
  v9[0] = vert_up[3]; v9[1] = vert_dn[1]; v9[2] = vert_dn[0];
  polygons[9] = new Polygon(v9,NULL);

  std::vector<Vertex*> v10(3);
  v10[0] = vert_dn[3]; v10[1] = vert_dn[2]; v10[2] = vert_up[2];
  polygons[10] = new Polygon(v10,NULL);

  std::vector<Vertex*> v11(3);
  v11[0] = vert_up[2]; v11[1] = vert_up[1]; v11[2] = vert_dn[3];
  polygons[11] = new Polygon(v11,NULL);

  CSG* mesh = CSG::fromPolygons(polygons);
  for(int i=0;i<refine;i++){
    mesh = mesh->refine();
  }
  return mesh;
}

File: pyg4ometry/geant4/solid/TwistedBox.py
from ... import config as _config

from .SolidBase import SolidBase as _SolidBase

if _config.meshing == _config.meshingType.pycsg :
    from pyg4ometry.pycsg.core import CSG as _CSG
    from pyg4ometry.pycsg.geom import Vector as _Vector
    from pyg4ometry.pycsg.geom import Vertex as _Vertex
    from pyg4ometry.pycsg.geom import Polygon as _Polygon
elif _config.meshing == _config.meshingType.cgal_sm :
    from pyg4ometry.pycgal.core import CSG as _CSG
    from pyg4ometry.pycgal.geom import Vector as _Vector
    from pyg4ometry.pycgal.geom import Vertex as _Vertex
    from pyg4ometry.pycgal.geom import Polygon as _Polygon

from .TwistedSolid import TwistedSolid as _TwistedSolid
from .TwoVector import TwoVector as _TwoVector
from .Layer import Layer as _Layer

import numpy as _np
import logging as _log

class TwistedBox(_SolidBase, _TwistedSolid):
    """
    Constructs a box that is twisted though angle twisted angle

    :param name:         of the solid
    :type name:          str
    :param twistedangle: twist angle, must be less than pi/2
    :type twistedangle:  float, Constant, Quantity, Variable, Expression
    :param pDx:          length in x
    :type pDx:           float, Constant, Quantity, Variable, Expression
    :param pDy:          length in y
    :type pDy:           float, Constant, Quantity, Variable, Expression
    :param pDz:          length in z
    :type pDz:           float, Constant, Quantity, Variable, Expression
    :param refine:       number of steps to iteratively smoothen the mesh by doubling the number of vertices at every step
    :type refine:        int
    :param registry:     for storing solid
    :type registry:      Registry
    :param lunit:        length unit (nm,um,mm,m,km) for solid
    :type lunit:         str    
    :param aunit:        angle unit (rad,deg) for solid
    :type aunit:         str
    :param nstack:       Not written
    :type nstack:        int
    """
    def __init__(self, name, twistedangle, pDx, pDy, pDz, registry,
                 lunit = "mm", aunit = "rad",
                 nstack=None, refine=0, addRegistry=True):
        super(TwistedBox, self).__init__(name, 'TwistedBox', registry)

        self.twistedAngle = twistedangle
        self.pDx          = pDx
        self.pDy          = pDy
        self.pDz          = pDz
        self.lunit        = lunit
        self.aunit        = aunit
        self.nstack       = nstack if nstack else _config.SolidDefaults.TwistedBox.nstack
        self.refine       = refine

        self.dependents = []

        self.varNames = ["twistedAngle", "pDx", "pDy", "pDz"]
        self.varUnits = ["aunit", "lunit", "lunit", "lunit"]

        self.checkParameters()

        if addRegistry:
            registry.addSolid(self)

    def __repr__(self):
        return "Twisted Box : {} {} {} {} {}".format(self.name, self.twistedAngle,
                                                     self.pDx, self.pDy, self.pDz)
    def checkParameters(self):
        if self.evaluateParameterWithUnits('twistedAngle') > _np.pi:
            raise ValueError("Twisted Angle must be less than 0.5*pi")


    def makeLayers(self, p1, p2, p3, p4, pDz, theta, nstack):
        dz = 2*pDz/nstack
        dtheta = theta/nstack
        z = -pDz

        layers = []

        bottom = _Layer(p1,p2,p3,p4, z)
        bottom = bottom.Rotated(-theta*0.5) #overwrite
        layers.append(bottom)

        for i in range(nstack):
            l = layers[-1].Rotated(dtheta) # returns rotated copy
            z += dz # increment z
            l.z = z # fix z
            layers.append(l)

        return layers

    def mesh_old(self):
        _log.info('twistedbox.pycsgmesh> antlr')

        import pyg4ometry.gdml.Units as _Units  # TODO move circular import
        luval = _Units.unit(self.lunit)
        auval = _Units.unit(self.aunit)

        twistedAngle = self.evaluateParameter(self.twistedAngle) * auval
        pDx = self.evaluateParameter(self.pDx) / 2. * luval
        pDy = self.evaluateParameter(self.pDy) / 2. * luval
        pDz = self.evaluateParameter(self.pDz) / 2. * luval
        refine = self.evaluateParameter(self.refine)

        p1 = _TwoVector(-pDx, -pDy)  # , pDz]
        p2 = _TwoVector(pDx, -pDy)  # pDz]
        p3 = _TwoVector(pDx, pDy)  # pDz]
        p4 = _TwoVector(-pDx, pDy)  # pDz]
        m = self.makeLayers(p1, p2, p3, p4, pDz, -twistedAngle, self.nstack)

        return self.meshFromLayers(m, self.nstack)

    def makeLayerEdge(self, pDx, pDy, pTwistedAngle = 0, nx = 20, ny = 20):
        x = []
        y = []

        dX = 2*pDx/nx
        dY = 2*pDy/ny

        # +y
        for i in range(0,ny+1) :
            x.append(-pDx)
            y.append(-pDy+i*dY)

        # +x
        for i in range(1,nx+1) :
            x.append(-pDx+i*dX)
            y.append(pDy)

        # -y
        for i in range(1,ny+1) :
            x.append(pDx)
            y.append(pDy-i*dY)

        # -x
        for i in range(1,nx) :
            x.append(pDx-i*dX)
            y.append(-pDy)

        x = _np.array(x)
        y = _np.array(y)

        nx = x*_np.cos(pTwistedAngle) - y*_np.sin(pTwistedAngle)
        ny = x*_np.sin(pTwistedAngle) + y*_np.cos(pTwistedAngle)

        x = list(nx)
        y = list(ny)

        #import matplotlib.pyplot as _plt
        #_plt.plot(x,y,"+")

        return [x,y]

    def mesh(self):
        _log.info('twistedbox.pycsgmesh> antlr')

        import pyg4ometry.gdml.Units as _Units #TODO move circular import 
        luval = _Units.unit(self.lunit)
        auval = _Units.unit(self.aunit) 

        twistedAngle = self.evaluateParameter(self.twistedAngle)*auval
        pDx = self.evaluateParameter(self.pDx)/2.*luval
        pDy = self.evaluateParameter(self.pDy)/2.*luval
        pDz = self.evaluateParameter(self.pDz)/2.*luval
        refine  = self.evaluateParameter(self.refine)

        _log.info('twistedbox.mesh> mesh')

        dTwist = twistedAngle/self.nstack
        dZ     = 2*pDz/self.nstack

        polygons = []

        for i in range(0,self.nstack,1) :

            i1 = i
            i2 = i+1

            twist1 = -twistedAngle/2.0 + i1*dTwist
            z1     = -pDz+i1*dZ

            twist2 = -twistedAngle/2.0 + i2*dTwist
            z2     = -pDz+i2*dZ

            edge1 = self.makeLayerEdge(pDx, pDy, twist1, 20, 20)
            edge2 = self.makeLayerEdge(pDx, pDy, twist2, 20, 20)

            # add bottom layer
            if i1 == 0 :
                vBottom = []
                for j in range(0,len(edge1[0]),1) :
                    vBottom.append(_Vertex(_Vector([edge1[0][j],edge1[1][j], z1])))
                polygons.append(_Polygon(vBottom))

            # add top layer
            if i2 == self.nstack :
                vTop = []
                for j in range(0, len(edge1[0]), 1):
                    vTop.append(_Vertex(_Vector([edge2[0][j], edge2[1][j], z2])))
                vTop.reverse()
                polygons.append(_Polygon(vTop))

            # edges
            for j in range(0,len(edge1[0]),1) :
                j1 = j
                j2 = (j+1) % len(edge1[0])
                vSide = []
                vSide.append(_Vertex(_Vector([edge1[0][j1],edge1[1][j1], z1])))
                vSide.append(_Vertex(_Vector([edge2[0][j1],edge2[1][j1], z2])))
                vSide.append(_Vertex(_Vector([edge2[0][j2],edge2[1][j2], z2])))
                vSide.append(_Vertex(_Vector([edge1[0][j2],edge1[1][j2], z1])))
                polygons.append(_Polygon(vSide))

            # print(i1,twist1,z1,i2,twist2,z2)


        return _CSG.fromPolygons(polygons)
